src/netlink_namespace.c

```diff
@@ -683,34 +682,59 @@ static int ir_config(const char *key, const char *value) {
 
 static int ir_init (void)
 {
-  DEBUG("Initializing netlink_namespace plugin\n");
-  if (ns_name != NULL) {
-    DEBUG("netlink_namespace plugin: ir_init: Switching to namespace on file '%s'.", *ns_name);
+  int selfns, targetns;
 
-    int fd = open(*ns_name, O_RDONLY);
-    if (fd  < 0) {
-        ERROR("Unable to open() %s", *ns_name);
-    }
-
-    if (setns(fd, CLONE_NEWNET) != 0) {
-       ERROR("netlink_namespace plugin: ir_init: Unable to setns() to %s.", *ns_name);
-       return (-1);
-    }
-  } else {
+  DEBUG("Initializing netlink_namespace plugin\n");
+  if (ns_name == NULL) {
     ERROR("netlink_namespace plugin: ir_init: Did not get the namespace name?");
     return (-1);
   }
 
-  nl = mnl_socket_open (NETLINK_ROUTE);
-  if (nl == NULL)  {
-    ERROR ("netlink_namespace plugin: ir_init: mnl_socket_open failed.");
+  /* Save "current" namespace status */
+  if (NULL == (nloriginal = mnl_socket_open(NETLINK_ROUTE))) {
+    ERROR("netlink_namespace plugin: ir_init: Unable to open mnl_socket for current namespace  (errno %d)", EXIT_FAILURE);
     return (-1);
   }
 
-  if (mnl_socket_bind (nl, 0, MNL_SOCKET_AUTOPID) < 0)  {
-    ERROR ("netlink_namespace plugin: ir_init: mnl_socket_bind failed.");
+  if (0 > mnl_socket_bind(nloriginal, 0, MNL_SOCKET_AUTOPID)) {
+    ERROR("netlink_namespace plugin: ir_init: Unable to bind to netlink for current namespace (errno %d)", EXIT_FAILURE);
     return (-1);
   }
+
+  if ((selfns = open(SELFNS, O_RDONLY)) < 0) {
+    ERROR("netlink_namespace plugin: ir_init: Could not open self namespace file at '%s'", SELFNS);
+    return (-1);
+  }
+
+  /* Save custom configured namespace status */
+
+  DEBUG("netlink_namespace plugin: ir_init: Switching to namespace on file '%s'.", ns_name);
+  if ((targetns = open(ns_name, O_RDONLY)) < 0) {
+    ERROR("Unable to open namespace file '%s'", ns_name);
+    return(-1);
+  }
+
+  if (setns(targetns, CLONE_NEWNET) != 0) {
+    ERROR("netlink_namespace plugin: ir_init: Unable to switch namespace to '%s'", ns_name);
+    return(-1);
+  }
+
+  if (NULL == (nltarget = mnl_socket_open(NETLINK_ROUTE))) {
+    ERROR("netlink_namespace plugin: ir_init: Unable to open socket to namespace '%s'", ns_name);
+    return(-1);
+  }
+
+  if (0 < mnl_socket_bind(nltarget, 0, MNL_SOCKET_AUTOPID)) {
+    ERROR("netlink_namespace plugin: ir_init: Unable to bind to target namespace");
+    return(-1);
+  }
+
+  /* Get back to main namespace */
+  if (setns(selfns, CLONE_NEWNET) != 0) {
+    ERROR("netlink_namespace plugin: ir_init: Unable to get back to original namespace");
+    return(-1);
+  }
+ 
   return (0);
```

Comment:
In the context of this particular use in collectd, this is a nitpick, in general it's not. Consider the lifetime of selfns and targetns.

- selfns/targetns is variables local to ir_init, you essentially leak them as soon as ir_init returns.
- You also on error return, without closing anything you may have opened as well.

What you could consider is what i.e. the linux kernel does:

```
int some_function() {
  int somefd, otherfd;

  if (0 > (somefd = open(something, O_RDONLY))) {
    ERROR("bad stuff");
    goto cleanup;
  }
  if (0 > (otherfd = open(something, O_RDONLY))) {
    ERROR("badness");
    goto cleanup;
  }

  return 0;
cleanup:
  if (somefd >= 0 && close(somefd) != 0) { ERROR("close failed: %m"); };
  if (otherfd >= 0 && close(otherfd) != 0) { ERROR("close failed: %m"); };
  return -1  
}
```

Why this could be bad - in this case, I'm relatively sure that ir_init is only called once, although imagine that collectd tried init multiple times, you'll now essentially every time ir_init is called leak some file-descriptors, at some point you'll come up against RLIMIT_NOFILE and everything will potentially break.